File: extras/Python/cmd_messenger/communication_manager.py
"""``CommunicationManager`` — port of C# ``CommunicationManager``.

Sits between the transport layer (raw bytes) and the receive command queue
(parsed :class:`ReceivedCommand` objects), and provides the synchronous
``execute_send_*`` methods used by both the send queue and the synchronous
:meth:`CmdMessenger.send_command_sync` path.
"""
from __future__ import annotations

import threading
from typing import TYPE_CHECKING, Optional

from . import escaping, time_utils
from .enums import BoardType, SendQueue
from .escaping import IsEscaped
from .received_command import ReceivedCommand
from .send_command import SendCommand
from .transport import Transport

if TYPE_CHECKING:
    from .queue.receive_command_queue import ReceiveCommandQueue


class CommunicationManager:
    """Manager for data over the transport layer."""

    def __init__(
        self,
        transport: Transport,
        receive_command_queue: "ReceiveCommandQueue",
        board_type: BoardType,
        command_separator: str,
        field_separator: str,
        escape_character: str,
    ) -> None:
        self._transport = transport
        self._receive_command_queue = receive_command_queue

        self.board_type: BoardType = board_type
        self.command_separator: str = command_separator
        self.field_separator: str = field_separator
        self.escape_character: str = escape_character

        #: Whether to append ``\r\n`` after every written command.
        self.print_lf_cr: bool = False
        #: Time-stamp (ms) of the most recently received line.
        self.last_line_time_stamp: int = 0

        self._send_command_data_lock = threading.RLock()
        self._parse_lines_lock = threading.Lock()
        self._buffer: str = ""
        self._is_escaped = IsEscaped()

        # Wire transport → parser.
        self._transport.data_received += self._new_data_received

    # ------------------------------------------------------------------
    # Connection
    # ------------------------------------------------------------------
    def connect(self) -> bool:
        if self._transport.is_connected():
            return False
        return self._transport.connect()

    def disconnect(self) -> bool:
        if not self._transport.is_connected():
            return False
        return self._transport.disconnect()

    # ------------------------------------------------------------------
    # Raw write
    # ------------------------------------------------------------------
    def write(self, value: str) -> None:
        """Write a string (ISO-8859-1 encoded) to the transport."""
        self._transport.write(value.encode("iso-8859-1"))

    def write_line(self, value: str) -> None:
        """Write a string followed by ``\\r\\n``."""
        self.write(value + "\r\n")

    # ------------------------------------------------------------------
    # Synchronous send paths used by SendCommandQueue
    # ------------------------------------------------------------------
    def execute_send_command(
        self, send_command: SendCommand, send_queue_state: SendQueue
    ) -> ReceivedCommand:
        """Send a command, optionally waiting for an acknowledge.

        Returns a :class:`ReceivedCommand`. The result is only meaningful when
        ``send_command.req_ac`` is true; otherwise an empty stub is returned.
        """
        with self._send_command_data_lock:
            send_command.communication_manager = self
            send_command.init_arguments()

            if send_command.req_ac:
                self._receive_command_queue.suspend()
                self._receive_command_queue.prepare_for_cmd(
                    send_command.ack_cmd_id, send_queue_state
                )
                self._write_command(send_command)
                rc = self._receive_command_queue.wait_for_cmd(send_command.timeout)
                ack_command = rc if rc is not None else ReceivedCommand()
            else:
                self._write_command(send_command)
                ack_command = ReceivedCommand()

            ack_command.communication_manager = self

        if send_command.req_ac:
            self._receive_command_queue.resume()

        return ack_command

    def execute_send_string(
        self, command_string: str, send_queue_state: SendQueue
    ) -> ReceivedCommand:
        """Send a pre-formatted command string to the transport."""
        with self._send_command_data_lock:
            if self.print_lf_cr:
                self.write_line(command_string)
            else:
                self.write(command_string)
        rc = ReceivedCommand()
        rc.communication_manager = self
        return rc

    # ------------------------------------------------------------------
    # Parsing pipeline
    # ------------------------------------------------------------------
    def _new_data_received(self, *_args, **_kwargs) -> None:
        self._parse_lines()

    def _parse_lines(self) -> None:
        with self._parse_lines_lock:
            data = self._transport.read()
            if data:
                self._buffer += data.decode("iso-8859-1")
            while True:
                current_line = self._parse_line()
                if not current_line:
                    break
                self.last_line_time_stamp = time_utils.millis()
                self._process_line(current_line)

    def _process_line(self, line: str) -> None:
        received = self._parse_message(line)
        received.raw_string = line
        received.time_stamp = self.last_line_time_stamp
        self._receive_command_queue.queue_received_command(received)

    def _parse_message(self, line: str) -> ReceivedCommand:
        cleaned = line.strip("\r\n")
        cleaned = escaping.remove(cleaned, self.command_separator, self.escape_character)
        args = escaping.split(
            cleaned,
            self.field_separator,
            self.escape_character,
            remove_empty_entries=True,
        )
        rc = ReceivedCommand(args)
        rc.communication_manager = self
        return rc

    def _parse_line(self) -> str:
        if not self._buffer:
            return ""
        i = self._find_next_eol()
        if 0 <= i < len(self._buffer):
            line = self._buffer[: i + 1]
            self._buffer = self._buffer[i + 1 :]
            return line
        return ""

    def _find_next_eol(self) -> int:
        pos = 0
        while pos < len(self._buffer):
            escaped = self._is_escaped.escaped_char(self._buffer[pos])
            if self._buffer[pos] == self.command_separator and not escaped:
                return pos
            pos += 1
        return pos

    def _write_command(self, send_command: SendCommand) -> None:
        if self.print_lf_cr:
            self.write_line(send_command.command_string())
        else:
            self.write(send_command.command_string())

    # ------------------------------------------------------------------
    # Disposal
    # ------------------------------------------------------------------
    def dispose(self) -> None:
        try:
            self._transport.data_received -= self._new_data_received
        except Exception:
            pass

    def __enter__(self) -> "CommunicationManager":
        return self

    def __exit__(self, *exc_info) -> None:
        self.dispose()

File: extras/TypeScript/src/communicationManager.ts
import { CommandContext } from './command.js';
import { latin1Decode, latin1Encode } from './encoding.js';
import { BoardType, SendQueue } from './enums.js';
import { IsEscaped, remove, split } from './escaping.js';
import { ReceiveCommandQueue } from './queue/receiveCommandQueue.js';
import { ReceivedCommand } from './receivedCommand.js';
import { SendCommand } from './sendCommand.js';
import { millis } from './timeUtils.js';
import { ITransport } from './transport/transport.js';

export class CommunicationManager implements CommandContext {
  printLfCr = false;
  lastLineTimeStamp = 0;
  private buffer = '';

  constructor(
    private readonly transport: ITransport,
    private readonly receiveCommandQueue: ReceiveCommandQueue,
    public readonly boardType: BoardType,
    public readonly commandSeparator = ';',
    public readonly fieldSeparator = ',',
    public readonly escapeCharacter = '/'
  ) {
    this.transport.dataReceived.on('data', this.newDataReceived);
  }

  async connect(): Promise<boolean> {
    if (this.transport.isConnected()) {
      return false;
    }
    return this.transport.connect();
  }

  async disconnect(): Promise<boolean> {
    if (!this.transport.isConnected()) {
      return false;
    }
    return this.transport.disconnect();
  }

  async write(value: string): Promise<void> {
    await this.transport.write(latin1Encode(value));
  }

  async writeLine(value: string): Promise<void> {
    await this.write(`${value}\r\n`);
  }

  async executeSendCommand(sendCommand: SendCommand, sendQueueState: SendQueue): Promise<ReceivedCommand> {
    sendCommand.communicationManager = this;
    sendCommand.initArguments();

    if (!sendCommand.reqAc) {
      await this.writeCommand(sendCommand);
      return this.emptyReceivedCommand();
    }

    this.receiveCommandQueue.suspend();
    try {
      this.receiveCommandQueue.prepareForCmd(sendCommand.ackCmdId, sendQueueState);
      await this.writeCommand(sendCommand);
      const received = await this.receiveCommandQueue.waitForCmd(sendCommand.timeout);
      const ack = received ?? new ReceivedCommand();
      ack.communicationManager = this;
      return ack;
    } finally {
      this.receiveCommandQueue.resume();
    }
  }

  async executeSendString(commandString: string, _sendQueueState: SendQueue): Promise<ReceivedCommand> {
    if (this.printLfCr) {
      await this.writeLine(commandString);
    } else {
      await this.write(commandString);
    }
    return this.emptyReceivedCommand();
  }

  dispose(): void {
    this.transport.dataReceived.off('data', this.newDataReceived);
  }

  private readonly newDataReceived = (): void => {
    this.parseLines();
  };

  private parseLines(): void {
    const data = this.transport.read();
    if (data.length > 0) {
      this.buffer += latin1Decode(data);
    }

    while (true) {
      const currentLine = this.parseLine();
      if (currentLine === '') {
        break;
      }
      this.lastLineTimeStamp = millis();
      this.processLine(currentLine);
    }
  }

  private parseLine(): string {
    if (this.buffer.length === 0) {
      return '';
    }

    const endOfLine = this.findNextEndOfLine();
    if (endOfLine >= 0) {
      const line = this.buffer.slice(0, endOfLine + 1);
      this.buffer = this.buffer.slice(endOfLine + 1);
      return line;
    }

    return '';
  }

  private findNextEndOfLine(): number {
    const escaped = new IsEscaped(this.escapeCharacter);
    for (let i = 0; i < this.buffer.length; i += 1) {
      const char = this.buffer[i] ?? '';
      const currentEscaped = escaped.escapedChar(char);
      if (char === this.commandSeparator && !currentEscaped) {
        return i;
      }
    }
    return -1;
  }

  private processLine(line: string): void {
    const received = this.parseMessage(line);
    received.rawString = line;
    received.timeStamp = this.lastLineTimeStamp;
    this.receiveCommandQueue.queueReceivedCommand(received);
  }

  private parseMessage(line: string): ReceivedCommand {
    const cleaned = remove(line.replace(/[\r\n]+$/u, ''), this.commandSeparator, this.escapeCharacter);
    const args = split(cleaned, this.fieldSeparator, this.escapeCharacter, true);
    const received = new ReceivedCommand(args);
    received.communicationManager = this;
    return received;
  }

  private async writeCommand(sendCommand: SendCommand): Promise<void> {
    if (this.printLfCr) {
      await this.writeLine(sendCommand.commandString());
    } else {
      await this.write(sendCommand.commandString());
    }
  }

  private emptyReceivedCommand(): ReceivedCommand {
    const command = new ReceivedCommand();
    command.communicationManager = this;
    return command;
  }
}
